File: No_1293/Shortest Path in a Grid with Obstacles Elimination.py
class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        m = len(grid)
        n = len(grid[0])
        
        dis = 0
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        visited = [[-1] * n for _ in range(m)]
        q = collections.deque()
        q.append(((0, 0), k))

        while q:
            count = len(q)
            while count:
                cell, old_k = q.popleft()
                i, j = cell
                count -= 1
                
                if i == m - 1 and j == n - 1:
                    return dis
                for dr, dc in directions:
                    new_i = i + dr
                    new_j = j + dc
                    if new_i < 0 or new_i >= m or new_j < 0 or new_j >= n:
                        continue
                        
                    new_k = old_k - grid[new_i][new_j]
                    if new_k < 0:
                        continue
                        
                    if visited[new_i][new_j] >= new_k:
                        continue

                    visited[new_i][new_j] = new_k

                    q.append(((new_i, new_j), new_k))
            dis += 1
            
        return -1

File: No_1293/Shortest Path in a Grid with Obstacles Elimination.cpp
class Solution {
public:
    int shortestPath(vector<vector<int>>& grid, int k) {
        int m = grid.size(), n = grid[0].size();
        
        int dis = 0;
        vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        vector<vector<int>> visited(m, vector<int>(n, -1));
        queue<pair<pair<int, int>, int>> q;
        q.push({{0, 0}, k});
        
        while (!q.empty()) {
            int size = q.size();
            while (size--) {
                auto & [cell, old_k] = q.front();
                auto & [i, j] = cell;
            
                if (i == m - 1 && j == n - 1)
                    return dis;
                
                for (auto & [dr, dc] : directions) {
                    int new_i = i + dr;
                    int new_j = j + dc;
                    if (new_i < 0 || new_i >= m || new_j < 0 || new_j >= n)
                        continue;
                    
                    int new_k = old_k - grid[new_i][new_j];
                    if (new_k < 0)
                        continue;
                    
                    if (visited[new_i][new_j] >= new_k)
                        continue;

                    visited[new_i][new_j] = new_k;
                    q.push({{new_i, new_j}, new_k});
                }
                q.pop();
            }
            dis++;
        }
        
        return -1;
    }
};